File: lecture_notes/10 refactoring/7-optimized.py
import ast

def shift(offset, points):
    for point in points:
        point['x'] += offset['x']
        point['y'] += offset['y']
    return points

def parsing(points):
    parsed = []
    for point in points:
        if type(point) == str:
            parsed_point = ast.literal_eval(point)
            parsed.append(parsed_point)
        else:
            parsed.append(point)
    return parsed


polyline = [{'x': 0, 'y': 0},
            {'x': 10, 'y': 10},
            '{ "x": 20, "y": 20 }',
            {'x': 30, 'y': 30}]

parsed_polyline = parsing(polyline)
shifted = shift({'x': 10, 'y': -5}, parsed_polyline)

print(shifted)

File: lecture_notes/10 refactoring/7-optimized.js
'use strict';

const shift = (offset) => (point) => {
  const shifted = {
    x: (point.x += offset.x),
    y: (point.y += offset.y),
  };
  return shifted;
};

const conditional = (item) => {
  if (typeof item === 'object') return item;
  return JSON.parse(item);
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const offset = shift({ x: 10, y: -5 });
const parsed = polyline.map(conditional);
const shifted = parsed.map(offset);
console.log({ shifted });
